### DISCUSSION TITLE
Replace high-resolution images in pdf

### DISCUSSION THREAD

**Discussion by alesky78** [LibrePDF/OpenPDF#985] [Q&A]
I have a pdf PDF contains high-resolution images and i would like to find a way to reduce the resolution of the
images inside.

Looking for a solution i found that in the original iText library there was a way to do it and it is also described in the old book iText in action.

this is the code snippet placed in the book

```java
PdfReader reader = new PdfReader(src);
int n = reader.getXrefSize();
PdfObject object;
PRStream stream;
``` 

start lo look all the PdfObject inside the pdf
```java
	// Look for image and manipulate image stream
		for (int i = 0; i < n; i++) {
			object = reader.getPdfObject(i);
			if (object == null || !object.isStream())
				continue;
			stream = (PRStream)object;
			PdfObject pdfsubtype = stream.get(PdfName.SUBTYPE);
``` 

and then if the image is found, manipulate the BufferedImage  inside the PdfImageObject 
```java
if (pdfsubtype != null && pdfsubtype.toString().equals(PdfName.IMAGE.toString())) {
				PdfImageObject image = new PdfImageObject(stream);
				BufferedImage bi = image.getBufferedImage();
				if (bi == null) continue;
				int width = (int)(bi.getWidth() * FACTOR);
				int height = (int)(bi.getHeight() * FACTOR);
				BufferedImage img = new BufferedImage(width, height, BufferedImage.TYPE_INT_RGB);
				AffineTransform at = AffineTransform.getScaleInstance(FACTOR, FACTOR);
				Graphics2D g = img.createGraphics();
				g.drawRenderedImage(bi, at);
				ByteArrayOutputStream imgBytes = new ByteArrayOutputStream();
				ImageIO.write(img, "JPG", imgBytes);
				//stream.clear();
				stream.setData(imgBytes.toByteArray(), false, PRStream.BEST_COMPRESSION);
				stream.put(PdfName.TYPE, PdfName.XOBJECT);
				stream.put(PdfName.SUBTYPE, PdfName.IMAGE);
				stream.put(key, value);
				stream.put(PdfName.FILTER, PdfName.DCTDECODE);
				stream.put(PdfName.WIDTH, new PdfNumber(width));
				stream.put(PdfName.HEIGHT, new PdfNumber(height));
				stream.put(PdfName.BITSPERCOMPONENT, new PdfNumber(8));
				stream.put(PdfName.COLORSPACE, PdfName.DEVICERGB);
			}
``` 

the idea is clear and simple to understand, when you find a stream if this is related to an image
you can create an instance of PdfImageObject over the stream that aloud you to manipulate the Buffere of the image

Unfortunately, the PdfImageObject class is no longer present in LibrePDF. 

Does anyone happen to have an idea on how it might be possible to manipulate images inside the pdf?

**Comment by alesky78** [comment]
Digging little bit looks like that this is a functionality introduced in iText 5
and this is why it is not present in OpenPDF

**Comment by andreasrosdalw** [comment]
Pull requests welcome

**Comment by alesky78** [comment]
The PdfImageObject is present in the OpenPDF5

https://github.com/LibrePDF/OpenPDF5/blob/develop/itext/src/main/java/com/itextpdf/text/pdf/parser/PdfImageObject.java

a first compiling exercise could be to integrate PdfImageObject  in OpenPDF and see what happen
